firedrake/mesh.py

```diff
@@ -594,10 +620,135 @@ def create_cell_node_list(self, global_numbering, fiat_element):
+        self._ds = ufl.Measure('exterior_facet', subdomain_data=weakref.ref(coordinates))
+        self._dS = ufl.Measure('interior_facet', subdomain_data=weakref.ref(coordinates))
+        if self.layers:
+            self._ds_t = ufl.Measure('exterior_facet_top', subdomain_data=weakref.ref(coordinates))
+            self._ds_b = ufl.Measure('exterior_facet_bottom', subdomain_data=weakref.ref(coordinates))
+            self._ds_v = ufl.Measure('exterior_facet_vert', subdomain_data=weakref.ref(coordinates))
+            self._dS_h = ufl.Measure('interior_facet_horiz', subdomain_data=weakref.ref(coordinates))
+            self._dS_v = ufl.Measure('interior_facet_vert', subdomain_data=weakref.ref(coordinates))
+        # Set the subdomain_data on all the default measures to this
+        # coordinate field.
+        # We don't set the domain on the measure since this causes
+        # an uncollectable reference in the global space (dx is
+        # global).  Furthermore, it's never used anyway.
+        for measure in [ufl.dx, ufl.ds, ufl.dS]:
+            measure._subdomain_data = weakref.ref(coordinates)
+        if self.layers:
+            for measure in [ufl.ds_t, ufl.ds_b, ufl.ds_v, ufl.dS_h, ufl.dS_v]:
+                measure._subdomain_data = weakref.ref(coordinates)
+
+    def init(self):
+        """Finish the initialisation of the mesh.  Most of the time
+        this is carried out automatically, however, in some cases (for
+        example accessing a property of the mesh directly after
+        constructing it) you need to call this manually."""
+        if hasattr(self, '_callback'):
+            self._callback(self)
+
+    @property
+    def topology(self):
+        """The underlying mesh topology object."""
+        return self._topology
+
+    @property
+    def topological(self):
+        """Alias of topology.
+
+        This is to ensure consistent naming for some multigrid codes."""
+        return self._topology
+
+    def ufl_domain(self):
+        self.init()
+        return self._ufl_domain
+
+    def ufl_cell(self):
+        """The UFL :class:`~ufl.cell.Cell` associated with the mesh."""
+        return self.ufl_domain().cell()
+
+    @utils.cached_property
+    def _coordinates_function(self):
+        """The :class:`.Function` containing the coordinates of this mesh."""
+        import firedrake.functionspace as functionspace
+        import firedrake.function as function
+        self.init()
+
+        coordinates_fs = self._coordinates.function_space()
+        V = functionspace.WithGeometry(coordinates_fs, self)
+        f = function.Function(V, val=self._coordinates)
+        return f
+
+    @property
+    def coordinates(self):
+        """The :class:`.Function` containing the coordinates of this mesh."""
+        return self._coordinates_function
+
+    @coordinates.setter
+    def coordinates(self, value):
+        raise AttributeError("You can no longer change the coordinates.")
```

Comment:
Yes. `utils.cached_property` is for caching, and the other property is used to signal a failure if the user tries to change the coordinates. (Which would silently "succeed" (fail) if there was just a `utils.cached_property`, but have no effect on anything whatsoever.)